atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java

```diff
@@ -79,8 +100,32 @@ public ListenableFuture<Map<Long, Long>> get(Iterable<Long> cells) {
                 MoreExecutors.directExecutor());
     }
 
+    private synchronized FollowUpAction touchAndReturn(CellInfo cellAndValue) {
+        if (acceptStagingReadsAsCommitted.get()) {
+            return FollowUpAction.PUT;
+        }
+        Cell cell = cellAndValue.cell();
+        byte[] actual = cellAndValue.value();
+        try {
+            store.checkAndTouch(cell, actual);
```

Comment:
Something I don't quite understand with the transaction v3 protocol in step 4.i is why the `CAS({V, STAGING}, {V, STAGING})` followed by `PUT({V, COMMITTED})` instead of just a single `CAS({V, STAGING}, {V, COMMITTED})`. If we're going to eat the overhead of CAS with Cassandra Paxos, shouldn't we check for staging and set committed? If the CAS succeeds we're done and if a racing worker has already committed, then there's a no reason to PUT.

https://github.com/palantir/atlasdb/blob/develop/docs/source/miscellaneous/doc/adr/0017-transactions3.md

Reply:
If we do a CAS from staging to committed we are still vulnerable to the same bug as before. We could be in this state:

1. V, Staging
2. W, Staging
3. Nothing

If we do a CAS (V, Staging) to (V, Committed) and it was only partially applied, and then we also do a CAS (W, Staging) to (W, Committed), then we could end up with

1. V, Committed 
2. W, Committed
3. Nothing

Then a read from nodes (1, 3) will return V, Committed, but a later read from a quorum that contains (W, Committed) will return (W, Committed)